sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,519 @@
+        {
+            _eventSource = logContent ? eventSource : null;
+            _maxLength = maxLength;
+            _cancellationToken = cancellationToken;
+        }
+
+        public async ValueTask LogAsync(string requestId, bool isError, Stream? stream, Encoding? textEncoding, bool async)
+        {
+            EventType eventType = ResponseOrError(isError);
+
+            if (stream == null || !IsEnabled(eventType))
+            {
+                return;
+            }
+
+            var bytes = await FormatAsync(stream, async).ConfigureAwait(false).EnsureCompleted(async);
+            Log(requestId, eventType, bytes, textEncoding);
+        }
+
+        public async ValueTask LogAsync(string requestId, BinaryContent? content, Encoding? textEncoding, bool async)
+        {
+            EventType eventType = EventType.Request;
+
+            if (content == null || !IsEnabled(eventType))
+            {
+                return;
+            }
+
+            var bytes = await FormatAsync(content, async).ConfigureAwait(false).EnsureCompleted(async);
+
+            Log(requestId, eventType, bytes, textEncoding);
+        }
+
+        public void Log(string requestId, bool isError, byte[] buffer, int offset, int length, Encoding? textEncoding, int? block = null)
+        {
+            EventType eventType = ResponseOrError(isError);
+
+            if (buffer == null || !IsEnabled(eventType))
+            {
+                return;
+            }
+
+            var logLength = Math.Min(length, _maxLength);
+
+            byte[] bytes;
+            if (length == logLength && offset == 0)
+            {
+                bytes = buffer;
+            }
+            else
+            {
+                bytes = new byte[logLength];
+                Array.Copy(buffer, offset, bytes, 0, logLength);
+            }
+
+            Log(requestId, eventType, bytes, textEncoding, block);
+        }
+
+        public bool IsEnabled(bool isError)
+        {
+            return IsEnabled(ResponseOrError(isError));
+        }
+
+        private bool IsEnabled(EventType errorResponse)
+        {
+            return _eventSource != null &&
+                   (_eventSource.IsEnabled(EventLevel.Informational, EventKeywords.All) ||
+                   (errorResponse == EventType.ErrorResponse && _eventSource.IsEnabled(EventLevel.Warning, EventKeywords.All)));
+        }
+
+        private async ValueTask<byte[]> FormatAsync(BinaryContent requestContent, bool async)
+        {
+            using var memoryStream = new MaxLengthStream(_maxLength);
+
+            if (async)
+            {
+                await requestContent.WriteToAsync(memoryStream, _cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                requestContent.WriteTo(memoryStream, _cancellationToken);
+            }
+
+            return memoryStream.ToArray();
```

Comment:
Is this just truncating the BinaryContent buffer to _maxLength?  This feels like an inefficient way to do this -- i.e. to allocate a new array (the buffer in MemoryStream), write into it, and then allocate another copy of the array (in ToArray).  I'm wondering if we can do it more efficiently?

Reply:
> Is this just truncating the BinaryContent buffer to _maxLength? 
Yes

Reply:
Maybe it's possible to use TryGetBuffer or do what you suggested below, copying directly into a byte[]. I'll experiment a bit. I left it as-is from Azure.Core on the first iteration to get all the tests green but I think there are definitely some improvements that can be made.

Reply:
moved this conversation to https://github.com/Azure/azure-sdk-for-net/pull/43510/files#r1578420575